beyond-all-reason/Beyond-All-Reason
# "ColourString" refactor

### Work done

* Add Spring.Utilities.Color module (common/springUtilities/color.lua).
* Refactor all colourNames, StringColour copies and similar into Spring.Color.
* Remove apparently useless "ends in 0" check and fix. **<- While I think this is safe please review throughly**
* Add Color.ToStringEx for alpha and outline formatting.

#### Test steps
- Checkout branch as BAR.sdd
- Open skirmish
- Check everything still works 

#### Motivation

Started working on this because I'm working on some more widgets that also use the "StringColour" functionality, noticed several copies of the same code in different widgets. Didn't want to add yet more copies.

### Considerations

While this targets mainly a refactor to avoid duplications, I discovered a few things while working on it, and included them in this PR (specially the 'ends in 0' check).

If it's deemed too dangerous or controversial I can prepare another PR with just the refactor and maybe ToStringEx, and leave the rest for future PR(s).

In any case, please *let me know what is wanted* and I will adapt the PR accordingly.

**1. Header lives in common/luaUtilities/color.lua**

*after review moved this to common/springUtilities/color.lua*

Maybe this is not the best place, also could be called colorformat instead, but this way it keeps the name open so maybe other common color utilities can be added in the future.

**2. What about luarules/colors.h.lua and luaui/Headers/colors.h.lua**

I saw [them](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Headers/colors.h.lua), but first, it's two copies of the same thing, and second, it's already including a lot of globals. That's why I didn't put new functionality there.

I think there would be more value to move the colors.h functionality into a module that's accessible through Spring namespace or just VFS.Includable. Anyways, for now I decided against doing any of that to somewhat limit the scope of this PR.

**3. Way to access it through Spring.Color**

*after review moved this to Spring.Utilities.Color*

I find this more practical, but we can just have users VFS.Including it and not include into Spring namespace.

**4. The "doesn't end in 0" check**

Everywhere seems to be checking for int stringified values to not end in 0, but IMO, this must be because of some old engine bug (maybe [this one?](https://github.com/beyond-all-reason/spring/commit/15fb446487ba07a2b7b138978d996792df8a9212)).

I have throughly investigated this, and didn't find any proofs warranting the check. Also did some tests setting lots of things to 000 and didn't see any problems.

**Note** i could be wrong and be some issue somewhere that needs this.

Still, looking at the code, \\000 text color is already used in several places, like [gui_vote_interface.lua](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_vote_interface.lua#L225), [gui_converter_usage.lua](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_converter_usage.lua#L87), [gui_advplayerslist.lua](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_advplayerslist.lua#L2891), [gui_info.lua](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_info.lua#L1128). [gui_gridmenu.lua](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_gridmenu.lua#L1803), [gui_top_bar.lua](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_top_bar.lua#L418). Those are critical widgets and already doing it, so I don't see why it can't just be done everywhere.

Still, if you think this is too dangerous, I can prepare the branch with just "> 0 check", some examples in the engine do this, like [gui_hilight_unit.lua](https://github.com/beyond-all-reason/spring/blob/master/cont/examples/Widgets/gui_hilight_unit.lua#L396). Still, I think this is also an old example and also not needed.

**5. Addition of ToStringEx**

This uses the extended formatting, allowing to format alpha and outline color.

While this isn't used in the codebase atm, I personally have a need to use it to control text alpha. It's a bit out of scope for this PR, but in the spirit of avoiding another PR in the close future I just included it here now.

**6. Color or Colour?**

I see the engine using Color, the same with opengl, while widgets use Colour in many places. I used Color for consistency with engine.

### References

Some references I investigated to try to find a rationale for the "can't end in 0" or "can't be 0".

* [glFont.cpp:SkipColorCodesAndNewLines](https://github.com/beyond-all-reason/spring/blob/master/rts/Rendering/Fonts/glFont.cpp#L250) indicator processing
* [glFont.h indicator codes](https://github.com/beyond-all-reason/spring/blob/master/rts/Rendering/Fonts/glFont.h#L110)
* [TextWrap.cpp:SkipColorCodes](https://github.com/beyond-all-reason/spring/blob/master/rts/Rendering/Fonts/TextWrap.cpp#L38) here colorcodes seem to be parsed
* [TextWrap.cpp:RemergeColorCodes](https://github.com/beyond-all-reason/spring/blob/master/rts/Rendering/Fonts/TextWrap.cpp#L524) reformat the colorcodes into string
* [Color.h](https://github.com/beyond-all-reason/spring/blob/master/rts/System/Color.h) SColor where colorcodes are stored
